SYCL Rewrites of inclusive_scan_by_segment and exclusive_scan_by_segment

mmichel11 opened the pull request:
# Original Description
To increase performance of segmented reduction and segmented scans, the SYCL backend implementations of the algorithms have been rewritten using the SYCL group and sub-group functions. 

The new algorithms operate in two primary phases:
1. Perform a segmented scan or reduction within each individual work-group with equal work assignment, storing partially completed scan / reduction values to global memory.
2. Apply the partial reduction / scan values from previous work-group(s) to the current.

# Update 1/18/2023
The scope of this PR has been reduced to only include the inclusive and exclusive segmented scan rewrites. The sycl rewrite for reduce by segment has been moved to a separate branch and will be worked on separately. This updated PR contains correctness fixes from its initial state along with expanded tests for the segmented scans.

## Algorithmic Descriptions
The segmented scans are implemented in a two kernel approach using the lower level SYCL libraries in contrast to the existing implementation utilizing internal parallel patterns. The segmented scans require the provided binary operator to be associative. Furthermore, the implementation is only invoked if the underlying datatype possesses a known identity element. Otherwise, the legacy implementation will be invoked.

The algorithms proceed as follows:
1. The first kernel is launched and each work group is assigned a range of elements to process equally divided among work items. Each work item is by default assigned four elements to process which has yielded the best performance empirically. Within a work group, the work items perform a segmented inclusive or exclusive scan writing the results to the global output buffer. The last work item outputs its final scan output to global memory along with a flag stating if the work group encountered a segment end.
2. The second kernel is launched upon the first kernel's completion. Each work group collectively searches the global segment end flag buffer within the range `[0, work group id - 1]` to find the work group with the closest id that contains a segment end while reducing the partial scan outputs in the range `[closest seg end id, work group id - 1]`. Once this prefix value is determined, work items will apply the prefix to the values stored in the global output buffer up until the first segment end in the work group (if it exists).

## Performance Results
Performance results show significant speedup over the existing segmented scan implementations for varying segment sizes. To obtain performance results of these new rewrites, please contact Matthew Michel for more details.

MikeDvorskiy commented:
I there any performance measurement after that re-writing?
It would be very interesting to have a look.

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                        else
+                            __partials_acc[__group_id] = __accumulator;
+
+                        bool __seg_present = static_cast<bool>(__closest_seg_id);
```
can we avoid `static_cast` by writing `bool __seg_present = __closest_seg_id != 0;` ?

mmichel11 replied:
Yes, I will make this change

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                        else
+                            __partials_acc[__group_id] = __accumulator;
+
+                        __flag_type __seg_present = static_cast<__flag_type>(__closest_seg_id);
```
can we avoid `static_cast` here ?

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+    void
+    sycl_exclusive_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values,
+                                   _Range3&& __out_values, _BinaryPredicate __binary_pred, _BinaryOperator __binary_op,
+                                   _T __init = _T{})
```
`_T __init = _T{}` adds additional requirement for type` _T `. Can we avoid that?

mmichel11 replied:
Yes, I will change this.

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+
+            __cgh.parallel_for(
+                sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+                    __val_type __accumulator = {};
```
Additional requirement (default constructor ) on  `__val_type`. Can we avoid that?

danhoeflinger replied:
Investigating this a bit more, this implementation not only requires a default constructor, but also requires it to be the identity for the `__binary_op` and `__val_type`.  We likely should specialize and have two internal implementations, one which takes full advantage of a `__known_identity<__binary_op, __val_type>`, and another which catches the general case without a known identity, the existing scan and reduce algorithms use this strategy.

danhoeflinger replied:
This has been resolved with known_identity special casing for the scan algorithm(s).

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+    {
+        if constexpr (__scan_type == scan_type::exclusive)
+        {
+            sycl_exclusive_scan_by_segment(::std::forward<_ExecutionPolicy>(__exec), ::std::forward<_Range1>(__keys),
```
The semantics of  the `sycl_exclusive_scan_by_segment `and `sycl_inclusive_scan_by_segment` are up to 99% the same.
It leads that implementations are the same too. And it is what we have in the changes... Can we keep just one scan_by_segment implementation with small customization in a couple of places for inclusive/exclusive case?

MikeDvorskiy commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+/*
+ *  Copyright (c) Intel Corporation
+ *
+ *  Copyright 2008-2013 NVIDIA Corporation
```
Is it correct with legal perspective?

mmichel11 replied:
The algorithm implemented is similar to the `reduce_by_key` and `scan_by_key` implementations in CUB and moderngpu. It is not a direct usage of their source code, but I included the NVIDIA copyright for this reason.

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+
+    sycl::group_barrier(__item.get_group());
+
+    for (int32_t __i = 1; __i < __wgroup_size; __i += __i)
```
May be better to use ::std::size_t type like the type of __wgroup_size ?

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+    _IdxType __first = 0;
+    __local_acc[__local_id] = __accumulator;
+
+    sycl::group_barrier(__item.get_group());
```
Am I right that we may use one barrier as first command inside loop instead of these two barriers?

SergeyKopienko replied:
I am simply tried to understand differences of this place with, for example, reduce_impl function, in which barrier setted inside loop.

mmichel11 replied:
Putting the barrier as the first line within the for loop does not guarantee synchronization after the last write. The read within the last accessor is not safe in this situation, so we would have to have barrier after the for loop and before the return statement. This behavior would be the same as what I have currently implemented.

SergeyKopienko replied:
ok

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+        using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using __flag_type = bool;
+
+        const int __n = __keys.size();
```
Why int? May be auto or std::size_t ?

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+            ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+        int32_t __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
```
Why int32_t ? I think that the size of __n_groups is not large then the size of __n

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+            int __local_id = __group.get_local_id();
+            int __global_id = __item.get_global_id();
+
+            int __start = __global_id * __vals_per_item;
```
For result of __global_id * __vals_per_item you use int here

SergeyKopienko replied:
ok

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+
+            // 2b. Perform a serial scan within the work item over assigned elements. Store partial
+            // reductions in work group local memory.
+            int32_t __start = __global_id * __vals_per_item;
```
For result of __global_id * __vals_per_item you use int32_t here, in previous calculation - int

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                        __val_type __agg_collector{};
+                        bool __first = true;
+
+                        int __start = __global_id * __vals_per_item;
```
Again int for __global_id * __vals_per_item

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                    // 1a. Perform a serial scan within the work item over assigned elements. Store partial
+                    // reductions in work item memory, and write the accumulated value and number of counted
+                    // segments into work group memory.
+                    int __start = __global_id * __vals_per_item;
```
Again int

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                    // Edge case for when the first element is the start of a new segment. We do 
+                    // not want to apply a null aggregate since there is no guarantee it is an identity
+                    if (__start < __n && __local_id != 0 && __keys[__start] != __keys[__start - 1])
+                        __max_end = __local_id;
```
You assign std::size_t to int. Is it correct?

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+
+                    for (int32_t __i = __start; __i < __end; ++__i)
+                    {
+                        if (__first)
```
I once read that inside loops it is better to compose a condition in such a way that a more frequent situation is performed in the if branch and not in the else branch. I am not sure, is it applicable there, but may be...

danhoeflinger replied:
This is now out of date, as we have removed this branching by explicitly requiring and using a known_identity for the by_segment rewrites.

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                        auto __global_id = __item.get_global_id();
+                        auto __local_id = __item.get_local_id();
+
+                        int __wg_agg_idx = __group_id - 1;
```
You assign here std::size_t to int. Is it correct?

mmichel11 replied:
Yes, in this case this correct since the group's id may be 0, so we must use a signed type.

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+                    __val_type __accumulator = {};
+
+                    auto __group = __item.get_group();
+                    int __global_id = __item.get_global_id();
```
Why int instead of std::size_t ?

mmichel11 replied:
Changed to std::size_t

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+                        __ag_exists = true;
+                        for (int32_t __i = __wg_agg_idx; __i >= 0; --__i)
+                        {
+                            const auto& __wg_aggregate = __partials_acc[__i];
```
Is it really required to save this and next values by const reference? May be better simple copy them into local const variable?

mmichel11 replied:
The type of `__partials` is unknown, so I think it may be best to save by const reference since we do not know if the copy is expensive. However, the next line accesses a boolean value, so using a copy into a local const variable I agree is best.

danhoeflinger replied:
This is now out of date, we are just using these accessor values directly in the following lines rather than grabbing a copy / reference and using that.

SergeyKopienko commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,506 @@
+
+                        // 2c. Second pass over the keys, reidentifying end segments and applying work group
+                        // aggregates if appropriate.
+                        ::std::size_t __local_min_key_idx = __n - 1;
```
May we can execute this line for case when __n == 0 then we will have an error, I think.

mmichel11 replied:
Yes, this would be an error. I will change this is to a signed type.

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+    __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+        ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+    int __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
```
I think that required to check all evaluations of __n_groups and __n_groups type: my assumption that int may be incorrect type for these cases.

danhoeflinger replied:
Changed __n_groups to ::std::size_t,  Thanks.

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+        __cgh.parallel_for(sycl::nd_range<1>{__n_groups * __wgroup_size, __wgroup_size}, [=](sycl::nd_item<1> __item) {
+            auto __group = __item.get_group();
+            auto __group_id = __group.get_group_id();
+            int __local_id = __group.get_local_id();
```
Why int ?

SergeyKopienko commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,288 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+            auto __group = __item.get_group();
+            auto __group_id = __group.get_group_id();
+            int __local_id = __group.get_local_id();
+            int __global_id = __item.get_global_id();
```
Why int ?

SergeyKopienko commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -44,7 +45,7 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
     const auto n = ::std::distance(first1, last1);
 
-    // Check for empty and single element ranges
+    // Check for empty element ranges
```
Why this comment was changed? It's describes two next conditions, I think

mmichel11 replied:
This was an unintended change that I did not realize was made. I will revert this back.

danhoeflinger commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+
+    const auto __n = __keys.size();
+
+    constexpr int __vals_per_item = 2; // Each work item serially processes 2 items. Best observed performance on gpu
```
Same comment as for scan.  Do we want to hardcode this? Does this need to be selected based on the device, or do we need to do a more general examination of a best global value here?

danhoeflinger commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+            auto __local_id = __group.get_local_id();
+            ::std::size_t __global_id = __item.get_global_id();
+
+            __val_type __accumulator{};
```
As mentioned below: extra requirement put on __val_type of default constructor.

danhoeflinger replied:
removed extra requirement of default constructor for __val_type.

danhoeflinger commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+            bool __first = true;
+            for (int32_t __i = __start; __i < __end; ++__i)
+            {
+                if (__first)
```
Can remove branch if `__accumulator` is set to identity for `__val_type` and `binary_op`.

danhoeflinger replied:
resolved.

danhoeflinger commented on include/oneapi/dpl/internal/reduce_by_segment_impl.h:
```diff
@@ -138,6 +138,293 @@ reduce_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 la
+
+            // 1a. Work item scan to identify segment ends
+            for (int32_t __i = __start; __i < __end; ++__i)
+                if (__n - 1 == __i || __keys[__i] != __keys[__i + 1])
```
For this and all other checks of equality of `__keys` and their neighbor, we need to use `__binary_pred()`.

danhoeflinger replied:
corrected.

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+                        int32_t __wg_agg_idx = __group_id - 1;
+                        __val_type __agg_collector{};
+                        
+                        // 2a. Calculate the work group's carry-in value.
```
This section (2a) uses thread 0 by itself to calculate the accumulated carry in value of all previous workgroups involved in the current segment.  For most segment sizes this is minimal work, which spans a small number of previous workgroups.

However, for workgroups beginning at the end of a very large segment that spans many workgroups, this can be a non-trivial amount of work that is done by a single thread, with the rest of the workgroup waiting.

The trouble is, we don't know how large the segment is, or how many previous workgroups it spans until we are done.  We need to decide if any special detection / handling of very large segments is necessary, it depends on our target use case.  

Its likely that any attempt to fix this will drag down the "normal" case.

danhoeflinger replied:
This was with a commit to make this section parallel (there still exists a small serial section, but it is acceptable).

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+        using __val_type = oneapi::dpl::__internal::__value_t<_Range2>;
+        using __flag_type = bool;
+
+        const int __n = __keys.size();
```
fixed this.

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(
+            ::std::forward<_ExecutionPolicy>(__exec), sizeof(__key_type) + sizeof(__val_type), __wgroup_size);
+
+        auto __n_groups = 1 + std::ceil(__n / (__wgroup_size * __vals_per_item));
```
corrected this

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+                    {
+                        // Edge case for when the first element is the start of a new segment. We do 
+                        // not want to apply a null aggregate since there is no guarantee it is an identity
+                        if (__start < __n && __local_id != 0 && __keys[__start] != __keys[__start - 1])
```
corrected the use of `__binary_pred`.

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+                                __local_min_key_idx, __dpl_sycl::__minimum<decltype(__local_min_key_idx)>());
+
+                        if (__group_id == 0)
+                            return;
```
moved this to cover all of 2c

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+                        }
+                    }
+
+                    __loc_acc[__local_id] = __accumulator;
```
removed the redundant line.

danhoeflinger commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,356 @@
+                                        const auto& __wg_aggregate = __partials_acc[__i];
+                                        const auto __b_seg_end = __seg_ends_acc[__i];
+
+                                        if (__first)
+                                        {
+                                            __agg_collector = __wg_aggregate;
+                                            __first = false;
+                                        }
+                                        else
+                                            __agg_collector = __binary_op(__wg_aggregate, __agg_collector);
+
```
We likely want to specialize for __known_identity<_BinaryOp, _Tp> as is done in reduce_impl.

See include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h:L245-L281 .

danhoeflinger replied:
This has been done for scan algorithms.

danhoeflinger commented:
Rebased to pick up bugfix for legacy range implementation.

SergeyKopienko commented on test/support/utils_test_base.h:
```diff
@@ -437,6 +451,48 @@ test_algo_three_sequences()
+        // create iterators
+        auto inout1_offset_first = test_base_data.get_start_from(UDTKind::eKeys);
+        auto inout2_offset_first = test_base_data.get_start_from(UDTKind::eVals);
+        auto inout3_offset_first = test_base_data.get_start_from(UDTKind::eResKeys);
```
May be better to use order as you describe in enum UDTKind ?

SergeyKopienko commented on test/support/utils_sycl.h:
```diff
@@ -300,6 +300,73 @@ test3buffers(int mult = kDefaultMultValue)
+        // 2. create iterators over buffers
+        auto inout1_offset_first = test_base_data.get_start_from(UDTKind::eKeys);
+        auto inout2_offset_first = test_base_data.get_start_from(UDTKind::eVals);
+        auto inout3_offset_first = test_base_data.get_start_from(UDTKind::eResKeys);
```
The same about order

julianmi commented on test/support/utils_sycl.h:
```diff
@@ -300,6 +300,73 @@ test3buffers(int mult = kDefaultMultValue)
+        // 2. create pointers at first+offset
+        auto inout1_offset_first = test_base_data.get_start_from(UDTKind::eKeys);
+        auto inout2_offset_first = test_base_data.get_start_from(UDTKind::eVals);
+        auto inout3_offset_first = test_base_data.get_start_from(UDTKind::eResKeys);
```
Thanks, I changed the order.

julianmi commented on include/oneapi/dpl/internal/inclusive_scan_by_segment_impl.h:
```diff
@@ -19,10 +19,10 @@
 #ifndef _ONEDPL_INCLUSIVE_SCAN_BY_SEGMENT_IMPL_H
 #define _ONEDPL_INCLUSIVE_SCAN_BY_SEGMENT_IMPL_H
 
+#include "by_segment_extension_defs.h"
```
I fixed the includes and the tests pass now.

danhoeflinger commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -95,18 +96,52 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
+    return result + n;
+}
+
+template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator, typename T,
+          typename BinaryPredicate, typename Operator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
+exclusive_scan_by_segment_impl_helper(Policy&& policy, InputIterator1 first1, InputIterator1 last1,
+                                      InputIterator2 first2, OutputIterator result, T init, BinaryPredicate binary_pred,
+                                      Operator binary_op, ::std::false_type /* has_known_identity*/)
+{
+
```
Currently, in this PR, we check if the binary_op and type have a "known identity" according to SYCL, and only use the rewrite implementation if they have a "known identity".  If they do not, we use the previous implementation.  

It's come to my attention that there is a known limitation which was lost from the documentation (and will be reintroduced) which requires that the default constructor for the scan algorithms be the identity for the type and binary_op.    In practice, I believe it is only truly required in a certain combination of compiler and backend (due to a requirement of openmp simd scan).  


I think we have two options here: 
1) Always use the new implementation of the scan_by_segment algorithms for SYCL supported backend, and define the identity to be the default constructor wherever SYCL known identity does not exist.  When we reintroduce the known limitation to the documentation, keep the language broad.
2) Use the previous implementation for cases which do not have a "known identity". When we reintroduce the known limitation to the documentation, adjust the language of the specify the limitation only in the limited scope where it functionally exists in the implementation.


Option (1) will result in faster performance for cases which do not have a sycl known identity. However, it has the possibility to break some existing user code which worked in previous versions.  (2) Will keep the same level of support as we currently provide for binary_op / type combinations where the default constructor is not the identity.  However, the cases in question are likely rare end-cases.

mmichel11 replied:
I think option 2 is the more conservative approach but is the current best option. For most use cases, I suppose that the sycl known identity will be present and the newer implementation would be invoked. Even if the cases in which existing user code could break are quite rare, I think this occurrence could cause quite some confusion since the limitation was not specified in the previous documentation.

rarutyun replied:
It's not a conservative. It's what we should do. Breaking the existing code is not an option. We (oneDPL) are generic library and we should care about our previous design and API even if we some internal optimization based on "something known :)" We could break something if we believe something is worth deprecating. But that's a separate discussion

SergeyKopienko commented:
@mmichel11 I think required to update this branch from current main branch state.

danhoeflinger commented on test/support/utils_test_base.h:
```diff
@@ -233,6 +236,7 @@ struct test_base_data_sequence : test_base_data<TestValueType>
     };
     ::std::vector<Data> data;   // Vector of source test data:
                                 //  - 3 items for test_algo_three_sequences
+                                //  - 4 items for test_algo_four_sequences
```
Good point.  When we reduced the scope of this PR to just scan algorithms, the usage was removed with reduce.  We should also remove these additions in this PR.

mmichel11 replied:
I have these currently unused tests within the branch where the segmented reduction implementation now lives, so I will remove these tests from this branch.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,335 @@
+
+                    __flag_type __group_has_segment_break = (__closest_seg_id != NO_SEGMENT_BREAK);
+
+                    __closest_seg_id = ::std::max(0L, __closest_seg_id); //get rid of no segment end found flag
```
I am assuming this is caused by the min / max macros in the Windows headers? I will replace this with `__dpl_sycl::__maximum` so it does not clash with these Windows macros.

danhoeflinger replied:
A `0L` (long literal) on windows is 4 bytes, and we were comparing to a explicitly 8 byte integer.   I've made it explicitly a `::std::int64_t(0)` to resolve it.

julianmi reviewed (commented):
I have only minor comments since I have worked with this code before.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+            ::std::forward<_ExecutionPolicy>(__exec), __kernel2) });
+#endif
+
+        ::std::size_t __n_groups = 1 + ((__n - 1) / (__wgroup_size * __vals_per_item));
```
I will use this function.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                    __flag_type __group_has_segment_break = (__closest_seg_id != NO_SEGMENT_BREAK);
+
+                    //get rid of no segment end found flag
+                    __closest_seg_id = ::std::max(::std::int64_t(0), __closest_seg_id);
```
Initially, we had just a literal type (0L) in here. However, on certain systems a long may only be 4 bytes. This will cause a compiler error here. I think it is safest to cast the value so we guarantee that the type of the 0 value is 64 bits.

danhoeflinger replied:
We should be able to use `0LL` I think and guarantee 64 bit on windows.  I'll take another look.

danhoeflinger replied:
I believe `::std::int64_t(0)` is the safe option.  `0LL` would work now, but there are no guarantees long long stays 64bit, only that its a minimum of 64bit.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                            __loc_seg_ends_acc[__local_id] = false;
+                            __loc_partials_acc[__local_id] = __identity;
+
+                            for (int64_t __i = __wg_agg_idx - __vals_to_explore * __local_id;;
```
Since we are counting downwards, it may be possible for a work item to receive a negative value of `__i`. This is checked in the nested loop below this line. We need `__i` to be a signed type for this reason which is why `int64_t` is used.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                                // serial aggregate collection and synchronization
+                                if (__local_id == 0)
+                                {
+                                    for (int64_t __j = 0; __j < __wgroup_size; ++__j)
```
Yes, although we just need to be careful that the datatype can hold the value of all possible maximum work group sizes across all devices. In general, is `uint16_t` something that we can safely use to iterate up to the maximum work group size for any device?

danhoeflinger replied:
As far as I can find there are no guarantees on sycl max work group sizes (other than their type being `size_t`).  If we put our own limits on the max work group size then we could limit this data type, but its only worthwhile if its a non-trivial performance benefit.

mmichel11 commented on test/support/utils_test_base.h:
```diff
@@ -36,7 +36,8 @@ enum class UDTKind
     eVals,      // ID for the second input data (if applicable)
-    eRes        // ID for the output data (result)
+    eRes,       // ID for the output data (result)
+    eRes2       // ID for the second output data (if applicable)
```
I removed this enum value.

mmichel11 commented on test/support/utils_const.h:
```diff
@@ -28,6 +28,7 @@ constexpr ::std::size_t max_n = 100000;
 constexpr ::std::size_t inout1_offset = 3;
 constexpr ::std::size_t inout2_offset = 5;
 constexpr ::std::size_t inout3_offset = 7;
+constexpr ::std::size_t inout4_offset = 9;
```
This const value has been removed.

mmichel11 commented on test/support/utils_test_base.h:
```diff
@@ -188,6 +190,7 @@ struct test_base_data_buffer : test_base_data<TestValueType>
                                 //  - 1 item for test1buffer;
                                 //  - 2 items for test2buffers;
                                 //  - 3 items for test3buffers
+                                //  - 4 items for test4buffers
```
This function has been removed.

mmichel11 commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,98 +16,116 @@
 // This macro may be used to analyze source data and test results in test_inclusive_scan_by_segment
 // WARNING: in the case of using this macro debug output is very large.
-//#define DUMP_CHECK_RESULTS
+// #define DUMP_CHECK_RESULTS
```
I reverted this change back.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,335 @@
+
+            oneapi::dpl::__ranges::__require_access(__cgh, __keys, __values, __out_values);
+
+            auto __loc_acc = sycl::accessor<__val_type, 1, sycl::access::mode::read_write, sycl::access::target::local>{
```
I have updated this (and all of the related instances) to use `__local_accessor`

mmichel11 commented on test/support/utils_test_base.h:
```diff
@@ -143,6 +144,7 @@ struct test_base_data_usm : test_base_data<TestValueType>
                                 //  - 1 item for test1buffer;
                                 //  - 2 items for test2buffers;
                                 //  - 3 items for test3buffers
+                                //  - 4 items for test4buffers
```
This function has been removed from the PR.

mmichel11 commented on test/support/utils_test_base.h:
```diff
@@ -437,6 +451,48 @@ test_algo_three_sequences()
+typename ::std::enable_if<
+    ::std::is_base_of<test_base<typename TestName::UsedValueType>, TestName>::value,
+    void>::type
+test_algo_four_sequences()
```
This function is not required for testing the segmented scans, so it has been removed from the PR.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+
+    __dpl_sycl::__group_barrier(__item);
+
+    for (::std::size_t __i = 1; __i < __wgroup_size; __i += __i)
```
`__i += __i` is the intended operation. I think changing this to `i *= 2` will make it more easily understood, so I will change this.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                    ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                    // First work item must set their accumulator to the provided init
+                    if (__global_id == 0 && __local_id == 0)
```
Yes, you are correct. I will remove this unnecessary check.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+
+        const ::std::size_t __n = __keys.size();
+
+        constexpr ::std::size_t __vals_per_item =
```
I agree that this is best to do. I will change it in all places.

mmichel11 replied:
I made this change for `__vals_per_item` since it will always be a small value. With regards to the type of `__wgroup_size`, I had a previous conversation with @danhoeflinger which concluded that it is best to leave it as is since the SYCL spec does not have a cap for the maximum work group size and returns a `std::size_t` when the work group size is queried. 

I know in the single work-group implementations, we can safely use `::std::uint16_t` to hold the size of the work group but do not believe this applies here. Do any other reviewers have any input on this? @adamfidel @SergeyKopienko @julianmi

julianmi replied:
I would use `std::size_t` here as well to align it with the SYCL spec.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+
+                    auto __group = __item.get_group();
+                    ::std::size_t __global_id = __item.get_global_id();
+                    ::std::size_t __local_id = __group.get_local_id();
```
I will make this change.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
+wg_segmented_scan(_NdItem __item, _LocalAcc __local_acc, _IdxType __local_id, _IdxType __delta_local_id,
```
Yes, I will add an explanation and example above the function definition.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                            __loc_seg_ends_acc[__local_id] = false;
+                            __loc_partials_acc[__local_id] = __identity;
+
+                            for (int64_t __i = __wg_agg_idx - __vals_to_explore * __local_id;;
```
The double `;;` is actually because the for loop test condition is omitted and the formatting makes it confusing. The for loop continues until a segment end has been identified and then breaks. I will look for a better way to structure this as I agree that it looks confusing.

The maximum value of `__i` is essentially the total number of work groups in the system. If we process `4` elements per work group and say we have a work group size of `1024`, then the total number of work groups hits `32,767` at an input size of `134,213,632`, and we would have an overflow with anything greater. I think it is reasonable to use `int32_t` instead, however.

mmichel11 replied:
I have reformatted this for loop to be more clear: the break condition is now checked in the test condition instead of the end of the loop. 

`__i` is now `::std::uint32_t` which is sufficiently large for all possible number of work groups.

adamfidel commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -95,18 +96,52 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
     if (n <= 0)
         return result;
 
+    typedef uint64_t CountType;
```
Use `::std::uint64_t` instead of `uint64_t`. Also, I would prefer `using` over `typedef`, but perhaps it's a personal choice.

mmichel11 replied:
I will make the change to `::std::uint64_t`. I normally use `using` as well, but used `typedef` to be consistent with the style of the already existing code in this header.

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+
+    __dpl_sycl::__group_barrier(__item);
+
+    for (::std::size_t __i = 1; __i < __wgroup_size; __i += __i)
```
You can change `__i` and `__wgroup_size` to be uint16_t if you follow Mikhail's suggestion to change the local ID type to 16-bits.

mmichel11 replied:
I will make this change.

mmichel11 replied:
After further discussion with others, it seems that it is best to leave the `__wgroup_size` as `std::size_t` to be consistent with the SYCL spec.

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+    sycl_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_values,
+                         _BinaryPredicate __binary_pred, _BinaryOperator __binary_op, _T __init, _T __identity)
+    {
+        using _Policy = typename ::std::decay<_ExecutionPolicy>::type;
```
This can be `::std::decay_t`.

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                        __accumulator = __init;
+                    }
+
+                    const int64_t NO_SEGMENT_BREAK = -1;
```
It probably doesn't matter, but this `const` can be `constexpr`. Also `int64_t` -> `::std::int64_t`.

mmichel11 replied:
I will make these changes.

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                    //get rid of no segment end found flag
+                    __closest_seg_id = ::std::max(::std::int64_t(0), __closest_seg_id);
+                    __val_type __carry_in = wg_segmented_scan(__item, __loc_acc, __local_id,
+                                                              __local_id - __closest_seg_id, __accumulator, __binary_op,
```
Should you cast `__local_id` to a signed type? Otherwise, would there be a risk of integer underflow here?

mmichel11 replied:
It is guaranteed that `__local_id >= __closest_seg_id` here since the search for the closest segment end is performed on lower indexed work items. However, it is possible that `__closest_seg_id` is -1. Is there any risk in subtracting -1 from an unsigned integer since the result is equivalent to adding one?

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                        if (__group_id != 0)
+                        {
+                            // 2a. Calculate the work group's carry-in value.
+                            bool __ag_exists = false;
+                            if (__start < __n)
+                            {
+                                __ag_exists = true;
+                            }
```
```suggestion
                            bool __ag_exists = __start < __n;
```

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                            }
+
+                            ::std::size_t __wg_min_seg_end = __dpl_sycl::__reduce_over_group(
+                                __group, __local_min_key_idx, __dpl_sycl::__minimum<decltype(__local_min_key_idx)>());
```
I think you can replace the `decltype` with `::std::size_t`, since you declared that variable just a few lines above.

mmichel11 replied:
Will make this change.

adamfidel commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+                            // local reductions followed by a downsweep
+                            // TODO: Generalize this value
+                            constexpr int64_t __vals_to_explore = 16;
+                            __flag_type __last_it = false;
```
I'm not sure I see the value of having a separate typedef for `__flag_type`, since it's always just `bool`.

mmichel11 replied:
I agree. I will change all instances of this to `bool` and remove `__flag_type`.

julianmi commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+        ::std::size_t __wgroup_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+
+        // change __wgroup_size according to local memory limit
+        __wgroup_size = oneapi::dpl::__internal::__max_local_allocation_size(::std::forward<_ExecutionPolicy>(__exec),
```
`__max_local_allocation_size` should be replaced with `__slm_adjusted_work_group_size` since #811.

mmichel11 replied:
I have updated this.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,372 @@
+ *  Copyright (c) Intel Corporation
+ *
+ *  Copyright 2008-2013 NVIDIA Corporation
+ *
```
Do any other reviewers have any insight on what should be changed in the comment header since Mikhail is OOO? It seems that we have two formats used in the codebase: this format which is seen in current mainline scan_by_segment, reduce_by_segment, and several other headers in this internal directory, and the format seen in the pstl directory and beyond.

In addition to the format itself being different, the format I am using lists the license as Apache 2.0 and the other header format uses Apache 2.0 with LLVM exception. Which is the correct to use here? @akukanov @rarutyun @SergeyKopienko

mmichel11 replied:
I have discussed with @ValentinaKats and have updated the header comment accordingly to provide the correct attributions.

mmichel11 commented:
I have addressed the comments on this PR thus far. I do have a remaining question with regards to the header file format Mikhail brought up which I have asked in response to his comment.

This branch has been rebased with the current main, and we are passing our updated tests across our configurations for the scan by segments. We should be in a good state to merge soon.

rarutyun commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -42,7 +43,7 @@ oneapi::dpl::__internal::__enable_if_host_execution_policy<typename ::std::decay
                                OutputIterator result, T init, BinaryPredicate binary_pred, Operator binary_op)
 {
-    const auto n = ::std::distance(first1, last1);
+    const auto n = last1 - first1;
```
why ``std::distance`` is "bad"?

MikeDvorskiy replied:
I think we should not cheat ourself with `distance`, because `exclusive_scan_by_segment_impl` accepts juts random access iterators. And of course it is better (and more transparency) to get compile time error on the first line of the implementation than somewhere in "the code details" below....

danhoeflinger replied:
If the motivation not to use `std::distance` is to use this to make a better compiler error when `InputIterator1` is not a random access iterator, should we instead require random access iterators in the function signature with an `__enable_if_host_execution_policy_conditional`?

mmichel11 replied:
Is the consensus to use `std::distance` with a static assert to ensure that only random access iterators are used?

mmichel11 replied:
I think because there are conflicting opinions on this issue it is just best to be consistent with the current code which uses `::std::distance`.

rarutyun commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -95,18 +96,50 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
+    auto buf_view = key_buf.all_view();
+    using iter_value_t = typename ::std::iterator_traits<InputIterator2>::value_type;
+
+    iter_value_t identity = unseq_backend::__known_identity<Operator, iter_value_t>;
```
Should it be `constexpr`?

mmichel11 replied:
Yes, I have made it `constexpr`.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,376 @@
+#    include <oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h>
+#    include <oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_utils.h>
+
+#    include <array>
```
what's for?

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,376 @@
+
+#if _ONEDPL_BACKEND_SYCL
+
+#    include <oneapi/dpl/pstl/algorithm_fwd.h>
```
Please shift all the includes left

mmichel11 replied:
I have made this change here and in other relevant places.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,376 @@
+    template <typename _ExecutionPolicy, typename _Range1, typename _Range2, typename _Range3,
+              typename _BinaryPredicate, typename _BinaryOperator, typename _T>
+    void
+    sycl_scan_by_segment(_ExecutionPolicy&& __exec, _Range1&& __keys, _Range2&& __values, _Range3&& __out_values,
```
What the point to have is as a separate method?

mmichel11 replied:
I have looked at this and agree it is unnecessary. I have removed this function and simply added the implementation in the function call operator.

rarutyun commented on include/oneapi/dpl/pstl/hetero/dpcpp/sycl_defs.h:
```diff
@@ -188,6 +188,39 @@ __reduce_over_group(_Args... __args)
+constexpr auto
+__any_of_group(_Args... __args)
+{
+#if _ONEDPL_SYCL2020_COLLECTIVES_PRESENT
```
Are we sure that either of APIs is available on all the compilers we are supporting today?

mmichel11 replied:
I will check on this and get back to you. These APIs are supported across our compiler configurations we use in CI, but I will verify this is the case for everything we support.

mmichel11 replied:
I have looked into this and the oneAPI group extensions for the algorithms I have added were provided in the `2021.1` DPC++ compiler alongside the scan and reduce group algorithms which we already use. The SYCL 2020 group algorithms were provided in `2021.4`. I believe it is therefore safe to add the rest of these group algorithms into `sycl_defs.h`.

rarutyun reviewed (commented):
This is partial review. I believe we need to have a call to discuss several things. Let us do it tomorrow (There are very small chances that I can make it today)

SergeyKopienko commented on test/support/utils_test_base.h:
```diff
@@ -373,6 +373,16 @@ void update_data(TTestDataTransfer& helper, Args&& ...args)
+    struct TestClassName : TestUtils::test_base<TestValueType>
+#else
+#define DEFINE_TEST_2(TestClassName, TemplateParams1, TemplateParams2)                      \
+    template <typename TemplateParams1, typename TemplateParams2>                           \
```
Please do not exclude first template param ```typename TestValueType``` in this case too:
- if we implement some class method outside the class, the declaration of class and template params should be the same.

SergeyKopienko replied:
So depends on ```#if TEST_DPCPP_BACKEND_PRESENT``` we may use inheritance `from ``TestUtils::test_base<TestValueType>``` or don't use it.

mmichel11 replied:
Maybe I am misunderstanding the suggestion, but it seems that `DEFINE_TEST_2` is written in alignment with the `DEFINE_TEST_1` and `DEFINE_TEST` macros above. The comment above these macros states to write them as:
```
// 1) define class as
//      template <typename TestValueType>
//      struct TestClassName : TestUtils::test_base<TestValueType>
// 2) define class as
//      struct TestClassName
```
Are you suggesting this structure be changed for `DEFINE_TEST_2`?

SergeyKopienko replied:
@mmichel11 you are right, it's ok.

SergeyKopienko commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,98 +16,116 @@
+        typedef typename ::std::iterator_traits<Iterator1>::value_type KeyT;
+        typedef typename ::std::iterator_traits<Iterator2>::value_type ValT;
+
+        ::std::string failure_msg = "wrong effect from inclusive_scan_by_segment with types [";
```
may be better to use ```std::string_stream```?

mmichel11 replied:
Yes, I agree since we are concatenating many strings. I used `std::ostringstream`.

SergeyKopienko replied:
Do you mean ```std::stringstream``` ?

mmichel11 replied:
Since we are using it for output only, I used [std::ostringstream](https://en.cppreference.com/w/cpp/io/basic_ostringstream) which is used in a few other places. It provides the same functionality as `std::stringstream` but only for output.

SergeyKopienko replied:
OK

SergeyKopienko commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -206,54 +223,86 @@ DEFINE_TEST_1(test_inclusive_scan_by_segment, BinaryOperation)
+    operator()(const _Tp& __x, const _Tp& __y) const
+    {
+        using KeyT = ::std::decay_t<decltype(__y)>;
+        return __y != KeyT(1);
```
May we use ```1```int the right part of comparison instead of ```KeyT(1)``` ?

mmichel11 replied:
This is okay for primitive datatypes but not okay for types such as `std::complex` where 1 cannot be converted to `std::complex`. For this reason it should be left as is.

SergeyKopienko replied:
ok

mmichel11 commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -206,54 +223,86 @@ DEFINE_TEST_1(test_inclusive_scan_by_segment, BinaryOperation)
+    bool
+    operator()(const _Tp& __x, const _Tp& __y) const
+    {
+        using KeyT = ::std::decay_t<decltype(__y)>;
```
Yes, I added this change

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+                        ::std::size_t __start = __global_id * __vals_per_item;
+                        ::std::size_t __end = __dpl_sycl::__minimum<decltype(__n)>{}(__start + __vals_per_item, __n);
+
+                        ::std::int32_t __wg_agg_idx = __group_id - 1;
```
is `std::int32t` sufficient? are we fine with potentially negative values?

mmichel11 replied:
We do not apply any aggregates with `__group_id` being 0, so the negative is disregarded in this case. The signed type is necessary to prevent an unsigned overflow below 0 for the arithmetic in the loop below.

I believe `int32_t` is sufficient since it stores the group ID meaning this would only cause an issue when we have over 2 billion+ work groups.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+                            bool __ag_exists = __start < __n;
+                            // local reductions followed by a downsweep
+                            // TODO: Generalize this value
+                            constexpr ::std::int64_t __vals_to_explore = 16;
```
Should it be ``std::int64_t`` or something narrower?

mmichel11 replied:
I have changed this to `std::int32_t`, so no additional conversions are made where it is used.

rarutyun commented on test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,73 +16,121 @@
-        while (i != n)
+        ::std::srand(42);
+        Size segment_length = 1;
+        for (Size i = 0; i != n;)
```
I understand that it was done before your changes but these nested loops look really weird

mmichel11 replied:
I have refactored this piece of code to use a single loop and be a bit more clear.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *  
+ * Copyright (c) 2013, NVIDIA CORPORATION.  All rights reserved.
```
I am worried about this licensing. Let's discuss why it's Nvidia stuff and probably rewrite it by our own

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+    return (__local_id ? __local_acc[__first + __local_id - 1] : __identity);
+}
+
+enum class scan_type
```
I am wondering why it's done via `enum` while further it's used as `bool` (if I am not missing something)

mmichel11 replied:
I just did this for stylistic reasons so its defined as either `sycl_scan_by_segment_impl<scan_type::inclusive> __scan` or `sycl_scan_by_segment_impl<scan_type::exclusive> __scan`.

I think changing it to a bool is more consistent with how we write  `sycl_scan_by_segment_impl</*is_inclusive=*/true> __scan` code in other areas, so I can change it to this.

mmichel11 replied:
I have made this change.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+
+#if _ONEDPL_BACKEND_SYCL
+
+#include <oneapi/dpl/pstl/algorithm_fwd.h>
```
you do have some `std` functionality there but don't have any includes

mmichel11 replied:
I see. I will add the includes for the `std` functionality we use in here.

rarutyun commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -152,9 +185,21 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
+
+template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator, typename T,
+          typename BinaryPredicate, typename Operator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
```
```suggestion
oneapi::dpl::__internal::__enable_if_hetero_execution_policy<::std::decay_t<Policy>, OutputIterator>
```

mmichel11 replied:
I have made this change.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+                        __accumulator = __init;
+                    }
+
+                    constexpr ::std::int64_t NO_SEGMENT_BREAK = -1;
```
Yes, I agree. I have made this change.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+                        __accumulator = __init;
+                    }
+
+                    constexpr ::std::int64_t NO_SEGMENT_BREAK = -1;
```
I have made this change.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+        using _CustomName = typename _Policy::kernel_name;
+
+        using _SegScanKernel1 = oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_generator<
+            _KernelName1, _CustomName, _Range1, _Range2, _Range3, _BinaryPredicate, _BinaryOperator>;
```
I have changed the naming to be descriptive of what the kernels actually do.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+            ::std::forward<_ExecutionPolicy>(__exec), 3 * sizeof(__val_type), __wgroup_size);
+
+#if _ONEDPL_COMPILE_KERNEL
+        auto __kernel1 = __par_backend_hetero::__internal::__kernel_compiler<_SegScanKernel1>::__compile(
```
I have made this change.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+
+                    constexpr ::std::int64_t NO_SEGMENT_BREAK = -1;
+                    // signed to allow flag for no segment break found
+                    ::std::int64_t __max_end = NO_SEGMENT_BREAK;
```
`__max_end` is just initialized to the value of `__no_segment_break` and may change later, so it cannot be `constexpr`. We need a signed type here, and I agree that `int64_t` is larger than necessary as `__max_end` stores the value of a local id (or the `no_segment_break` value). We had a separate discussion on this PR before and decided that we should not use `std::int16_t` to store work item local IDs, so I have changed to using `std::int32_t`.

mmichel11 commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,374 @@
+
+        auto __partials =
+            oneapi::dpl::__par_backend_hetero::__internal::__buffer<_ExecutionPolicy, __val_type>(__exec, __n_groups)
+                .get_buffer();
```
No, I do not believe there is a dangling risk here.

rarutyun commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -91,18 +92,50 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
     if (n <= 0)
         return result;
 
+    namespace __bknd = oneapi::dpl::__par_backend_hetero;
```
I think both exclusive and inclusive helpers can be combined in one function with extra parameter that indicates whether the scan is inclusive.

I believe there is no other difference between the code snippets other than exclusive vs inclusive. Please confirm

mmichel11 replied:
I can confirm this is the case. We still need unique implementations for the no identity case, so I will add a combined function in `scan_by_segment_impl.h` for when a known identity is present and takes inclusive / exclusive as a parameter

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,376 @@
+#if _ONEDPL_BACKEND_SYCL
+
+#include <type_traits>
+#include <cstdint>
```
Seems like `#include <cstddef>` is missed (for `std::size_t`)

mmichel11 replied:
Done.

MikeDvorskiy commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,98 +16,120 @@
+
+        ::std::ostringstream failure_msg;
+        failure_msg << "wrong effect from inclusive_scan_by_segment with types [";
+        failure_msg << typeid(KeyT).name();
```
Do we relay need a kind of RTTI here? (`typeid`)

mmichel11 replied:
I will remove this. I do not think it is necessary for the tests now since the branch is stable.

MikeDvorskiy commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,98 +16,120 @@
+        failure_msg << typeid(ValT).name();
+        failure_msg << "], ";
+        failure_msg << "n=" << ::std::to_string(n) << ", ";
+        if (::std::is_same<BinaryPredicateCheck, oneapi::dpl::__internal::__pstl_equal>::value)
```
`if constexpr` ?

mmichel11 replied:
Yes, I have added this.

rarutyun commented on include/oneapi/dpl/internal/inclusive_scan_by_segment_impl.h:
```diff
@@ -106,9 +122,21 @@ inclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
+template <typename Policy, typename InputIterator1, typename InputIterator2, typename OutputIterator,
+          typename BinaryPredicate, typename BinaryOperator>
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
+inclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIterator1 last1, InputIterator2 first2,
```
It seems like it could be pattern_nclusive_scan_by_segment and then the next level of abstraction is inclusive_scan_by_segment_impl and then scan_by_segement_impl_common. The same is true for exclusive one

mmichel11 replied:
I have made this change.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,419 @@
+                        __accumulator = __init;
+                    }
+
+                    constexpr ::std::int32_t __no_segment_break = ~0;
```
```suggestion
                    constexpr ::std::int32_t __no_segment_break = -1;
```
My bad, I thought `__no_segment_break` was `unsigned`

mmichel11 replied:
I have made this change and comment to use an `enum` in the future.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,419 @@
+
+    if (is_inclusive_scan)
+    {
+        __sycl_scan_by_segment_impl</*__is_inclusive=*/true>()(::std::forward<Policy>(policy), key_buf.all_view(),
```
```suggestion
        __sycl_scan_by_segment_impl<IsInclusive::value>()(::std::forward<Policy>(policy), key_buf.all_view(),
```

mmichel11 replied:
I have made these changes.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,419 @@
+oneapi::dpl::__internal::__enable_if_hetero_execution_policy<typename ::std::decay<Policy>::type, OutputIterator>
+__sycl_scan_by_segment_impl_invoker(Policy&& policy, InputIterator1 first1, InputIterator1 last1, InputIterator2 first2,
+                                    OutputIterator result, T init, BinaryPredicate binary_pred, Operator binary_op,
+                                    IsInclusive is_inclusive_scan)
```
```suggestion
                                    Inclusive)
```

mmichel11 replied:
Done

rarutyun commented on test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,73 +16,120 @@
 
 #if TEST_DPCPP_BACKEND_PRESENT
-#include "support/utils_sycl.h"
+#    include "support/utils_sycl.h"
```
```suggestion
#include "support/utils_sycl.h"
```

mmichel11 replied:
Done

rarutyun commented on test/parallel_api/numeric/numeric.ops/exclusive_scan_by_segment.pass.cpp:
```diff
@@ -200,22 +256,86 @@ DEFINE_TEST(test_exclusive_scan_by_segment)
+};
+
+template <typename _Tp>
+struct UserBinaryOperation<::std::complex<_Tp>>
```
Please move those predicates to some common place

mmichel11 replied:
I have moved these to a new header called `utils_functors.h` in the support directory to store functors such as this.

rarutyun commented on test/parallel_api/numeric/numeric.ops/inclusive_scan_by_segment.pass.cpp:
```diff
@@ -16,98 +16,116 @@
+        typedef typename ::std::iterator_traits<Iterator1>::value_type KeyT;
+        typedef typename ::std::iterator_traits<Iterator2>::value_type ValT;
+
+        ::std::ostringstream failure_msg;
```
Seems like a thing that was useful during a development but not right now. If you agree with that, could we just remove it? (and in other place as well)

mmichel11 replied:
Yes, I agree and have removed it.

rarutyun commented on include/oneapi/dpl/internal/exclusive_scan_by_segment_impl.h:
```diff
@@ -91,18 +92,27 @@ exclusive_scan_by_segment_impl(Policy&& policy, InputIterator1 first1, InputIter
 }
 
 #if _ONEDPL_BACKEND_SYCL
+
```
Is that something clang-format is suggesting? or is that just redundant change?

mmichel11 replied:
It was an unintended change. I will remove this empty line.

rarutyun commented on include/oneapi/dpl/internal/scan_by_segment_impl.h:
```diff
@@ -0,0 +1,408 @@
+// -----------------------------
+// __wg_segmented_scan is a derivative work and the reason for the additional copyright notice.
+template <typename _NdItem, typename _LocalAcc, typename _IdxType, typename _ValueType, typename _BinaryOp>
+inline _ValueType
```
Don't understand why it's inline

mmichel11 replied:
I agree there is no need for this to be inline. I have removed it.

rarutyun commented on test/support/utils_functors.h:
```diff
@@ -0,0 +1,63 @@
+// An arbitrary binary predicate to simulate a predicate the user providing
+// a custom predicate.
+template <typename _Tp>
+struct UserBinaryPredicate
```
Do those functors deserve separate header?

mmichel11 replied:
No, I have added these to `utils.h` and removed the new header.
